internal/providers/cloudstack/cloudstack.go

```diff
@@ -81,6 +86,11 @@ func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
 
 	<-ctx.Done()
 	if ctx.Err() == context.DeadlineExceeded {
+		// Did we hit neednet? If so, propagate that up instead. The OS should
+		// retry fetching again once networking is up.
+		if sawErrNeedNet {
```

Comment:
I have a feeling that this check needs to go between the channel-reading and the deadline-checking.

Reply:
Can you expand? Do you mean

```diff
diff --git a/internal/providers/openstack/openstack.go b/internal/providers/openstack/openstack.go
index d511bda2..5db67e52 100644
--- a/internal/providers/openstack/openstack.go
+++ b/internal/providers/openstack/openstack.go
@@ -85,6 +90,11 @@ func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
        })

        <-ctx.Done()
+       // Did we hit neednet? If so, propagate that up instead. The OS should
+       // retry fetching again once networking is up.
+       if sawErrNeedNet {
+               return types.Config{}, report.Report{}, resource.ErrNeedNet
+       }
        if ctx.Err() == context.DeadlineExceeded {
                f.Logger.Info("neither config drive nor metadata service were available in time. Continuing without a config...")
        }
```

instead?

If we did that, then on config-drive based OpenStack, we'd incorrectly return `ErrNeedNet`.

Reply:
Yes, that was the suggestion. But this parallel code is quite hard to reason about, so I'll wait for your manual tests.

Reply:
> If we did that, then on config-drive based OpenStack, we'd incorrectly return ErrNeedNet.

I read this a few times and now agree with this.